libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,383 @@
+		return true;
+	if (_op1.location == Location::Storage)
+	{
+		if (_op1.start && _op2.start)
+		{
+			yulAssert(
+				_op1.length &&
+				_op2.length &&
+				knowledge.valueIfKnownConstant(*_op1.length) == 1 &&
+				knowledge.valueIfKnownConstant(*_op2.length) == 1
+			);
+			return knowledge.knownToBeDifferent(*_op1.start, *_op2.start);
+		}
+	}
+	else
+	{
+		yulAssert(_op1.location == Location::Memory, "");
+		if (
+			(_op1.length && knowledge.knownToBeZero(*_op1.length)) ||
+			(_op2.length && knowledge.knownToBeZero(*_op2.length))
+		)
+			return true;
+
+		if (_op1.start && _op1.length && _op2.start)
+		{
+			optional<u256> length1 = knowledge.valueIfKnownConstant(*_op1.length);
+			optional<u256> start1 = knowledge.valueIfKnownConstant(*_op1.start);
+			optional<u256> start2 = knowledge.valueIfKnownConstant(*_op2.start);
+			if (
+				(length1 && start1 && start2) &&
+				*start1 + *length1 >= *start1 && // no overflow
+				*start1 + *length1 <= *start2
+			)
+				return true;
+		}
+		if (_op2.start && _op2.length && _op1.start)
+		{
+			optional<u256> length2 = knowledge.valueIfKnownConstant(*_op2.length);
+			optional<u256> start2 = knowledge.valueIfKnownConstant(*_op2.start);
+			optional<u256> start1 = knowledge.valueIfKnownConstant(*_op1.start);
+			if (
+				(length2 && start2 && start1) &&
+				*start2 + *length2 >= *start2 && // no overflow
+				*start2 + *length2 <= *start1
+			)
+				return true;
+		}
+
+		if (_op1.start && _op1.length && _op2.start && _op2.length)
+		{
+			optional<u256> length1 = knowledge.valueIfKnownConstant(*_op1.length);
+			optional<u256> length2 = knowledge.valueIfKnownConstant(*_op2.length);
+			if (
+				(length1 && *length1 <= 32) &&
+				(length2 && *length2 <= 32) &&
+				knowledge.knownToBeDifferentByAtLeast32(*_op1.start, *_op2.start)
+			)
+				return true;
+		}
+	}
+
+	return false;
+}
+
+bool UnusedStoreEliminator::knownCovered(
+	UnusedStoreEliminator::Operation const& _covered,
+	UnusedStoreEliminator::Operation const& _covering
+) const
+{
+	if (_covered.location != _covering.location)
+		return false;
+	if (
+		(_covered.start && _covered.start == _covering.start) &&
+		(_covered.length && _covered.length == _covering.length)
+	)
+		return true;
+	if (_covered.location == Location::Memory)
+	{
+		KnowledgeBase knowledge(m_dialect, m_ssaValues);
+
+		if (_covered.length && knowledge.knownToBeZero(*_covered.length))
+			return true;
+
+		// Condition (i = cover_i_ng, e = cover_e_d):
+		// i.start <= e.start && e.start + e.length <= i.start + i.length
+		if (!_covered.start || !_covering.start || !_covered.length || !_covering.length)
+			return false;
+		optional<u256> coveredLength = knowledge.valueIfKnownConstant(*_covered.length);
+		optional<u256> coveringLength = knowledge.valueIfKnownConstant(*_covering.length);
+		if (knowledge.knownToBeEqual(*_covered.start, *_covering.start))
+		{
+			if (coveredLength && coveringLength && *coveredLength <= *coveringLength)
+				return true;
+		}
```

Comment:
```suggestion
			if (coveredLength && coveringLength && *coveredLength <= *coveringLength)
				return true;
```
or continued ``&&``.